react/__tests__/OrderItems.test.tsx

```diff
@@ -0,0 +1,129 @@
+    const oldConsoleError = console.error
+    console.error = () => {}
+
+    const Component: FunctionComponent = () => {
+      useOrderItems()
+      return <div>foo</div>
+    }
+
+    expect(() =>
+      render(<Component />, {
+        graphql: { mocks: [ItemListMock] },
+      })
+    ).toThrow('useOrderItems must be used within a OrderItemsProvider')
+
+    console.error = oldConsoleError
+  })
+
+  it('should return a list of items when useOrderItems is called within a OrderItemsProvider', async () => {
+    const InnerComponent: FunctionComponent = () => {
+      const { itemList } = useOrderItems()
+      return (
+        <div>
+          {itemList.map((item: Item) => (
+            <div key={item.name}>{item.name}</div>
+          ))}
+        </div>
+      )
+    }
+
+    const OuterComponent: FunctionComponent = () => (
+      <OrderItemsProvider>
+        <InnerComponent />
+      </OrderItemsProvider>
+    )
+
+    const { getByText } = render(<OuterComponent />, {
+      graphql: { mocks: [ItemListMock] },
+    })
+    await new Promise(resolve => setTimeout(() => resolve())) // waits for graphql query
+
+    expect(getByText(mockItems[0].name)).toBeTruthy()
+    expect(getByText(mockItems[1].name)).toBeTruthy()
+    expect(getByText(mockItems[2].name)).toBeTruthy()
+  })
+
+  it('should optimistically update itemList when updateItem is called', async () => {
+    const oldConsoleError = console.error
+    console.error = () => {}
+
+    const UpdateItemMock = {
+      request: {
+        query: UpdateItem,
+        variables: {
+          orderItems: [
+            {
+              index: 1,
+              quantity: 123,
+            },
+          ],
+        },
+      },
+      result: {
+        data: {
+          updateItems: {
+            items: adjust(1, item => ({ ...item, quantity: 42 }), mockItems),
+          },
+        },
+      },
+    }
+
+    const InnerComponent: FunctionComponent = () => {
+      const { itemList, updateItem } = useOrderItems()
+      return (
+        <div>
+          {itemList.map((item: Item) => (
+            <div key={item.name}>
+              {item.name}: {item.quantity}
+            </div>
+          ))}
+          <button onClick={() => updateItem(1, 123)}>mutate</button>
+        </div>
+      )
+    }
+
+    const OuterComponent: FunctionComponent = () => (
+      <OrderItemsProvider>
+        <InnerComponent />
+      </OrderItemsProvider>
+    )
+
+    const { getByText } = render(<OuterComponent />, {
+      graphql: { mocks: [ItemListMock, UpdateItemMock] },
+    })
+    await new Promise(resolve => setTimeout(() => resolve())) // waits for initial item query
+
+    const button = getByText('mutate')
+    fireEvent.click(button)
+    expect(getByText(`${mockItems[1].name}: 123`)).toBeTruthy() // optimistic response
+
+    await new Promise(resolve => setTimeout(() => resolve())) // waits for actual mutation result
```

Comment:
We created the flushPromises API in test-tools, check it out

Reply:
😮